Firestore/core/test/firebase/firestore/util/filesystem_test.cc

```diff
@@ -179,6 +209,46 @@ TEST(FilesystemTest, RecursivelyDelete) {
   EXPECT_OK(RecursivelyDelete(nested_file));
 }
 
+TEST(FilesystemTest, RecursivelyDeleteTree) {
+  Path root_dir = Path::JoinUtf8(TempDir(), TestFilename());
+  Path middle_dir = Path::JoinUtf8(root_dir, "middle");
+  Path leaf1_dir = Path::JoinUtf8(middle_dir, "leaf1");
+  Path leaf2_dir = Path::JoinUtf8(middle_dir, "leaf2");
+  ASSERT_OK(RecursivelyCreateDir(leaf1_dir));
+  ASSERT_OK(RecursivelyCreateDir(leaf2_dir));
+
+  Touch(Path::JoinUtf8(middle_dir, "a"));
+  Touch(Path::JoinUtf8(middle_dir, "b"));
+  Touch(Path::JoinUtf8(leaf1_dir, "1"));
+  Touch(Path::JoinUtf8(leaf1_dir, "2"));
+  Touch(Path::JoinUtf8(leaf2_dir, "A"));
+  Touch(Path::JoinUtf8(leaf2_dir, "B"));
+  Touch(Path::JoinUtf8(leaf2_dir, "C"));
+
+  EXPECT_OK(RecursivelyDelete(root_dir));
+  EXPECT_NOT_FOUND(IsDirectory(root_dir));
+  EXPECT_NOT_FOUND(IsDirectory(leaf1_dir));
+}
+
+TEST(FilesystemTest, RecursivelyDeletePreservesPeers) {
+  Path root_dir = Path::JoinUtf8(TempDir(), TestFilename());
+
+  // Ensure that when deleting a directory we don't delete any directory that
+  // has a name that's a suffix of that directory. (This matters because on
+  // Win32 directories are traversed with a glob which can easily over-match.)
+  Path child = Path::JoinUtf8(root_dir, "child");
+  Path child_suffix = Path::JoinUtf8(root_dir, "child_suffix");
+
+  ASSERT_OK(RecursivelyCreateDir(child));
+  ASSERT_OK(RecursivelyCreateDir(child_suffix));
+  ASSERT_OK(IsDirectory(child_suffix));
```

Comment:
Strictly speaking, yes, though you can only establish that by looking at the other tests.

Since we're specifically testing that `RecursivelyDelete` does not harm this file during delete, it makes a test failure easier to understand if we assert the precondition here. I can't come up with a good way for that to fail though--I did this on principle more than anything. If you have any strong feelings about this I'll remove it.

Reply:
That was the principle I was operating under, but we've already tested that IsDirectory is a post-condition of RecursivelyCreateDir and this got up in your craw so I've deleted it.